### PR TITLE
Oceanwater 530 enable switching between arm controllers at runtime

### PR DESCRIPTION
**Pull request by Samahu**
## Linked Issues:
Linked PR: https://github.com/nasa/ow_autonomy/pull/15
Epic Link: https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-518
Linked Issues:
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-530
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-536
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-537
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-483 (api enhacements: remove delete_prev_traj)
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-526 (remove dependence on trajectory recording)

## Summary of Changes
* Added back the grinder_controller (deactivated by default)
* Created a method that enables switching between two sets of controllers at runtime
* Successfully ported the remaining services to using _JointTrajectoryController_:
  - **DigCircular**
  - **DigLinear**
  - **Grind**
* Removed `delete_prev_traj` from these services definitions as part of [OCEANWATER-483](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-483)
* Removed deprecated code that performs trajectory recording as part of [OCEANWATER-526](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-526)

```bash
roslaunch ow (atacama_y1a.launch | europa_terminator.launch | europa_terminator_workspace.launch)
```
Then either rqt_service_caller interface or launch a separate terminal (with the workspace sourced) and invoked any of the ported commands
```bash
rosservice call /arm/dig_linear ...
rosservice call /arm/dig_circular ...
rosservice call /arm/grind ...
```
### Test Objectives
* Make sure all services (newly ported and previous ones) are performing as expected.
* Note any improvements that can be made within the scope of this and linked PR.

### CODE REVIEW

File: ow_lander/launch/ros_controllers.launch [hunk 1]
```diff
@@ -5,10 +5,13 @@
...
-  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
-    output="screen" args="arm_controller antenna_controller"/>
-
   <node name="joint_controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
 		output="screen" args="joint_state_controller" />
 
+  <node name="controller_spawner_active" pkg="controller_manager" type="spawner" respawn="false"
+    output="screen" args="arm_controller antenna_controller"/>
+
+  <node name="controller_spawner_stopped" pkg="controller_manager" type="spawner" respawn="false"
```
Comment (mogumbo) [inline ow_lander/launch/ros_controllers.launch:14]: It's unclear to me what each controller is for based on the names and arguments. Some descriptive comments would be good.

File: ow_lander/scripts/path_planning_commander.py [hunk 2]
```diff
@@ -30,166 +29,129 @@ class PathPlanningCommander(object):
...
-    # move_grinder = moveit_commander.MoveGroupCommander("grinder")
-    display_trajectory_publisher = rospy.Publisher('/move_group/display_planned_path',
-                                                   moveit_msgs.msg.DisplayTrajectory,
-                                                   queue_size=20)
-    self.arm_move_group = move_arm
-    # self.move_limbs = move_limbs
-    # self.move_grinder = move_grinder
+
+    self.arm_move_group = moveit_commander.MoveGroupCommander("arm")
+    self.limbs_move_group = moveit_commander.MoveGroupCommander("limbs")
```
Comment (mogumbo) [inline ow_lander/scripts/path_planning_commander.py:34]: What does "limbs" refer to? Should be there an antenna group?

File: ow_lander/scripts/path_planning_commander.py [hunk 3]
```diff
@@ -30,166 +29,129 @@ class PathPlanningCommander(object):
...
+                                              moveit_msgs.msg.DisplayTrajectory,
+                                              queue_size=20)
 
   # === SERVICE ACTIVITIES - Stow =============================
-  def handle_stow(self, req): 
+  def handle_stow(self, req):
     """
     :type req: class 'ow_lander.srv._Stow.StowRequest'
     """
+    print("Unstow arm activity started")
```
Comment (mogumbo) [inline ow_lander/scripts/path_planning_commander.py:45]: Aren't "stow" and "unstow" swapped in these 4 print commands?

**Review by mogumbo — COMMENTED** [review]
I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.

File: ow_lander/scripts/path_planning_commander.py (hunk 3 above)
Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py:45]: Yes, you are right. I guess this is what reviews are for :sweat_smile:
I will address it

File: ow_lander/launch/ros_controllers.launch (hunk 1 above)
Comment (Samahu) [reply ow_lander/launch/ros_controllers.launch:16]: Addressed

File: ow_lander/scripts/path_planning_commander.py (hunk 2 above)
Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py:28]: The limbs group is same as the 'arm' group except it doesn't include the scope.

Good point on the antenna. The antenna handling was not part of this file but maybe it should .. I need to check where and how the antenna service is implemented (I will add a task for this to the epic).

File: ow_lander/scripts/path_planning_commander.py (hunk 3 above)
Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py]: addressed

**Comment by Samahu** [comment]
> I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.

With regard to the dig_circular doing the scooping motion high, your observation is right .. it is indeed higher than what we currently have on melodic-devel. I think part of the sequence of this activity is somehow being ignored. I will look into that.

File: ow_lander/scripts/path_planning_commander.py [hunk 4]
```diff
@@ -30,166 +23,129 @@ class PathPlanningCommander(object):
...
+    print("Grinde arm activity started")
+    grind_args = activity_grind.arg_parsing(req)
+    success = self.switch_controllers('grinder_controller', 'arm_controller')
+    if not success:
+      return False, "Failed"
+    success = activity_grind.grind(
+      self.grinder_move_group,
+      grind_args)
+    self.switch_controllers('arm_controller', 'grinder_controller')
+    print("Grinde arm activity completed")
```
Comment (damianacatanoso) [inline ow_lander/scripts/path_planning_commander.py:134]: Grind*

**Comment by Samahu** [comment]
> > I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.
> 
> With regard to the dig_circular doing the scooping motion high, your observation is right .. it is indeed higher than what we currently have on melodic-devel. I think part of the sequence of this activity is somehow being ignored which results in this different behavior. I will look into that.

BTW, I did check the issue about this circular movement and it stems from the fact the limbs_move_group commander is not able to run on the arm_controller (different set of joints). We need to add a new controller specific to the 'limbs move group' which is well understood. However, to resolve this problem for the dig_circular (and possibly the dig_linear too) activities I need to inject switch_controller commands (between arm_controller and the new limbs_controller) in the same manner I am doing it for the grind activity. That's definitely doable I just need to refactor these activities a bit.

**Comment by damianacatanoso** [comment]
> I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.
image
Same here. In my case there's an error (see screenshot). The error says: 
[ERROR] [1607731094.096570590, 1916525104.605000019]: Joints on incoming goal don't match the controller joints.
[ WARN] [1607731094.096696773, 1916525104.605000019]: Controller arm_controller failed with error INVALID_JOINTS: 
[ WARN] [1607731094.096734943, 1916525104.605000019]: Controller handle arm_controller reports status FAILED
[ INFO] [1607731094.096760379, 1916525104.605000019]: Completed trajectory execution with status FAILED ...
[ INFO] [1607731094.109268088, 1916525104.617500067]: ABORTED: Solution found but controller failed during execution
[ INFO] [1607731094.109377361, 1916525104.617500067]: Received event 'stop'

This means that the trajectory that takes the scoop down to the terrain, before the circular motion, never gets executed.

**Comment by Samahu** [comment]
> > I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.
> 
> image
> Same here. In my case there's an error (see screenshot). The error says:
> [ERROR] [1607731094.096570590, 1916525104.605000019]: Joints on incoming goal don't match the controller joints.
> [ WARN] [1607731094.096696773, 1916525104.605000019]: Controller arm_controller failed with error INVALID_JOINTS:
> [ WARN] [1607731094.096734943, 1916525104.605000019]: Controller handle arm_controller reports status FAILED
> [ INFO] [1607731094.096760379, 1916525104.605000019]: Completed trajectory execution with status FAILED ...
> [ INFO] [1607731094.109268088, 1916525104.617500067]: ABORTED: Solution found but controller failed during execution
> [ INFO] [1607731094.109377361, 1916525104.617500067]: Received event 'stop'
> 
> This means that the trajectory that takes the scoop down to the terrain, before the circular motion, never gets executed.

You are right, that what I was referring to in my previous comment. It is probably happening on the dig_linear too.

**Comment by damianacatanoso** [comment]
> > > I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.
> > 
> > 
> > With regard to the dig_circular doing the scooping motion high, your observation is right .. it is indeed higher than what we currently have on melodic-devel. I think part of the sequence of this activity is somehow being ignored which results in this different behavior. I will look into that.
> 
> BTW, I did check the issue about this circular movement and it stems from the fact the limbs_move_group commander is not able to run on the arm_controller (different set of joints). We need to add a new controller specific to the 'limbs move group' which is well understood. However, to resolve this problem for the dig_circular (and possibly the dig_linear too) activities I need to inject switch_controller commands (between arm_controller and the new limbs_controller) in the same manner I am doing it for the grind activity. That's definitely doable I just need to refactor these activities a bit.

move_limbs and move_arm are two separate group. move_arm moves the scoop, move_limbs moves the hand. Some of the scripts use move_limb so, yes, we need a separate controller for that. I can explain more in a call if you wish.

**Comment by damianacatanoso** [comment]
> > > > I ran all 3 services you suggested with defaults = true. dig_circular makes its scooping motion much too high. Other than that, everything looks alright.
> > > 
> > > 
> > > With regard to the dig_circular doing the scooping motion high, your observation is right .. it is indeed higher than what we currently have on melodic-devel. I think part of the sequence of this activity is somehow being ignored which results in this different behavior. I will look into that.
> > 
> > 
> > BTW, I did check the issue about this circular movement and it stems from the fact the limbs_move_group commander is not able to run on the arm_controller (different set of joints). We need to add a new controller specific to the 'limbs move group' which is well understood. However, to resolve this problem for the dig_circular (and possibly the dig_linear too) activities I need to inject switch_controller commands (between arm_controller and the new limbs_controller) in the same manner I am doing it for the grind activity. That's definitely doable I just need to refactor these activities a bit.
> 
> move_limbs and move_arm are two separate group. move_arm moves the scoop, move_limbs moves the hand. Some of the scripts use move_limb so, yes, we need a separate controller for that. I can explain more in a call if you wish.

**Comment by damianacatanoso** [comment]
I am so sorry, I closed the pull request by mistake! I reopened it!

**Review by mogumbo — APPROVED** [review]

**Review by damianacatanoso — APPROVED** [review]

**Pull request merged**